Make `FeeRate` use MvB internally

tcharding opened the pull request:
The `FeeRate` is a bit entangled with amount and weight.  Also we have an off-by-one bug caused by rounding errors and the fact that we use kwu internally.

We can get more precision in the fee rate by internally using per million virtual bytes.

- Fix: #4516
- Fix: #4497
- Fix: #3806

tcharding commented:
Question, do the checked functions in `fee.rs` _need_ to be const?

apoelstra commented:
> Question, do the checked functions in `fee.rs` _need_ to be const?

I'd like to get there eventually. But I think we'd be fine to release without it.

apoelstra commented:
The first two commits fa69cdec2fab63c9e207ca7a715b81c23ed80e4a and e38b5efb9de5869ab8748f32ab99494e08e80efd are unobjectionable and can be PR'd independently to shrink this one. The next on removes an `encapsulate` module so I guess needs to be in the big PR.

apoelstra commented on bitcoin/src/blockdata/script/borrowed.rs:
```diff
@@ -290,7 +290,7 @@ crate::internal_macros::define_extension_trait! {
         /// [`minimal_non_dust`]: Script::minimal_non_dust
         fn minimal_non_dust_custom(&self, dust_relay_fee: FeeRate) -> Option<Amount> {
-            self.minimal_non_dust_internal(dust_relay_fee.to_sat_per_kwu() * 4)
+            self.minimal_non_dust_internal(dust_relay_fee.to_sat_per_kwu_floor() * 4)
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

This one should be `_ceil` since we're trying to put a floor on the set of feerates that'll get broadcast.

BTW the commit message says that you add FIXME comments but actually you don't.

apoelstra commented on units/src/fee.rs:
```diff
@@ -80,7 +80,7 @@ impl Amount {
     pub const fn checked_div_by_fee_rate_floor(self, fee_rate: FeeRate) -> Option<Weight> {
         match self.to_sat().checked_mul(1000) {
-            Some(amount_msats) => match amount_msats.checked_div(fee_rate.to_sat_per_kwu()) {
+            Some(amount_msats) => match amount_msats.checked_div(fee_rate.to_sat_per_kwu_floor()) {
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

This one should be `_ceil` (somewhat weirdly, since the function name has `_floor` in it, but that's the nature of division).

tcharding replied:
Mad. Thanks for putting in the thought required to review this PR.

apoelstra commented on units/src/fee.rs:
```diff
@@ -151,7 +151,7 @@ impl FeeRate {
     /// Returns [`NumOpResult::Error`] if overflow occurred.
     pub const fn checked_mul_by_weight(self, weight: Weight) -> NumOpResult<Amount> {
-        if let Some(fee) = self.to_sat_per_kwu().checked_mul(weight.to_wu()) {
+        if let Some(fee) = self.to_sat_per_kwu_floor().checked_mul(weight.to_wu()) {
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

This one should be ceil.

apoelstra commented on units/src/fee_rate/mod.rs:
```diff
@@ -64,15 +64,18 @@ impl FeeRate {
     /// Converts to sat/vB rounding up.
     pub const fn to_sat_per_vb_ceil(self) -> u64 {
-        (self.to_sat_per_kwu() + (1000 / 4 - 1)) / (1000 / 4)
+        (self.to_sat_per_kwu_floor() + (1000 / 4 - 1)) / (1000 / 4)
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

This one should be `ceil`.

apoelstra commented on units/src/fee_rate/mod.rs:
```diff
@@ -81,7 +84,7 @@ impl FeeRate {
     #[must_use]
     pub const fn checked_mul(self, rhs: u64) -> Option<Self> {
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

I'm gonna OK these for now, but they need to be changed to do some sort of infallible conversion. Arguably they should be moved inside the `encapsulate` boundary.

apoelstra replied:
Same with `Sum`.

apoelstra replied:
And the `NonZeroU64` ops.

apoelstra commented on units/src/fee_rate/mod.rs:
```diff
@@ -117,34 +120,34 @@ impl FeeRate {
     }
 }
 
 impl From<FeeRate> for u64 {
```
In 4831689f0181c458dc6acf118414f56fcfe56bd1:

Can we just delete this impl? Not to channel Kix too hard but it's really unclear what it should mean.

apoelstra replied:
Ah, done in 95755218353f84b71fb3457ab8f1a9fff796e72d

apoelstra commented:
Done reviewing everything except 3603b50b2bdd023761dec982dcde5755a319b968 (the one that makes the actual change).

I identified 6 out of these 17 commits which are independent and unobjectionable. Can you PR those separately so we can reduce the size of this PR?

You can take or leave my comments about changing `_floor`s to `_ceil`s. I see that you replace many of them of with mvb conversions anyway.

apoelstra commented:
Also would be fine to PR separately (or alongside the other unobjectionable commits) to remove `const`s. We should file an issue to put them back. It's not hard, just annoying.

tcharding commented:
> are unobjectionable and can be PR'd independently to shrink this one

PR storm incomiiiiiiiiiiiiiing!

jrakibi commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -1657,9 +1653,9 @@ mod tests {
+            effective_value(fee_rate, InputWeightPrediction::P2WPKH_MAX, value);
 
         // 10 sat/kwu * 272 wu = 4 sats (rounding up)
         let expected_fee = "3 sats".parse::<SignedAmount>().unwrap();
```
I think this should be 3 sats, not 4 in the comment
```// 10 sat/kwu * 272 wu = 3 sats```

tcharding replied:
Ah yes, good eye. Thanks.

jrakibi commented on units/src/fee_rate/mod.rs:
```diff
@@ -36,50 +34,83 @@ impl FeeRate {
     /// 0 sat/kwu.
     ///
     /// Equivalent to [`MIN`](Self::MIN), may better express intent in some contexts.
-    pub const ZERO: FeeRate = FeeRate::from_sat_per_kwu(0);
+    pub const ZERO: FeeRate = FeeRate::from_sat_per_mvb(0);
```
Should we update the doc here to say `0 sat/MvB` to match the code implementation?
(Same thing for `const MIN` as well)

tcharding replied:
0 is 0. I have a feeling we should not mention the unit in docs at all unless its a function that specifically states the unit. I was going to go back over the docs again later, at a minimum the unit is mentioned, on `master` at least, on the `FeeRate` type itself.

apoelstra replied:
Agreed "zero is zero". It's sometimes clearer to not specify any unit at all to remind the reader that 0 is a real quantity, not an engineering convention. But if we do, it doesn't matter which one.

tcharding commented:
Thanks for the revew @jrakibi! Super useful to have extra eyes.

Kixunil commented:
Before we do this, is there any precedent in Bitcoin regarding this? Is any other library/wallet/... using this? If not why?

apoelstra commented:
@Kixunil because it means that for any transaction that fits in a block, we can go from a fee to a feerate and back without losing a single satoshi (if we're careful).

The alternative is to freeze on some arbitrary precision level, and then when our reference points (Core I guess) change we will just have to live with our fee computations being incorrect and theirs being correct. The difference will be on the order of single satoshis, which is kinda "whatever", but since we're aiming for a long-term 1.0 release I would rather we just start by doing something that's the best possible without introducing fraction types and/or arbitrary precision arithmetic.

apoelstra commented:
I would also be ok with just doing the API changes needed to support "go to MvB internally" but not actually doing it until we get bug reports about the precision. That would be:

* Splitting the sat/kwu conversions into _floor and _ceil variants (I think it's fine if these are unergonomic since most users will use the saturating `to_fee` method which just "figures this out", or just wants to display the rate which we will also abstract away).
* Lowering `FeeRate::MAX` by a factor of at least 1000.

yancyribbens commented on units/src/fee.rs:
```diff
@@ -51,20 +56,22 @@ impl Amount {
-                        return Some(FeeRate::from_sat_per_kwu(fee_rate));
-                    }
-                }
+            // No need to used checked arithmetic because wu is non-zero.
```
nit: this comment belongs to the line `let fee_rate = bump / wu;` which is one line lower.

apoelstra commented:
I don't think anybody will ever actually use "sat per MvB" or care that this is our internal representation.

If we went to a fractional type we would have to think about how to implement addition; exact fractional addition requires bignums and if we don't support exact addition then it seems like we've just added a bunch of complexity vs "use sat / MvB" to get the exact same properties.

yancyribbens commented:
Yes I see the constructors are still sat per kwu and vb so I see your point that external users of the API should never need to think in therms of MvB.  The downside looks like you can only represent smaller numbers due to checked mul by 1,000.  I expect the reasoning is that it's still plenty big enough for any real-world usage.

I haven't even thought about trying to implement FeeFrac and expect it would be a much more involved change as you point out.

tcharding commented:
The API is intrinsically tied to the inner representation. For example trying to add `to_sat_per_kvb_floor` requires overflow protection if inner rep is kwu (because we mul by 4). But if inner rep is MvB it does not require overflow protection. And the other `to_sat_per` functions do not have overflow protection so return `u64`.

I am pretty keen on changing the inner representation. So far no one has give any technically valid reasons why not to. "its different to Core" is not a valid technical reason.

tcharding commented:
Please note when reviewing https://github.com/rust-bitcoin/rust-bitcoin/pull/4534/commits/fad0a9b7da3d8d63c6615c7b5b71c5626ba4b995 - the ceil/floor calls may be confusing for that patch because the patch relies on the fact that the PR later changes the inner representation. To convince yourself please check the final state of the PR.

tcharding commented:
I tried to investigate just changing the API and not the inner rep but IMO it does not work. Force push is a minor docs typo fix and rebase.

Kixunil commented on bitcoin/src/blockdata/transaction.rs:
```diff
@@ -775,19 +774,15 @@ impl Decodable for Transaction {
+    let fee = fee_rate.to_fee(weight);
 
-    fee_rate.to_fee(weight).map(Amount::to_signed).and_then(|fee| value.to_signed() - fee) // Cannot overflow.
+    (value.to_signed() - fee.to_signed()).expect("cannot overflow")
```
Then you can't differentiate between overflow and just very low negative value rendering the negative values meaningless. Maybe it's not a big deal in practice but frankly, I don't like it.

yancyribbens replied:
> Anyway "cannot overflow" is a terrible comment; it took me way longer than it should to understand why.

I had a  more detailed explanation in the comment but was talked into something less verbose.  I wish there was a way in code to just not do a check since the check is expensive.  But it's not performance critical for my purpose.

Kixunil replied:
Technically, we can avoid the check if we want to, but I want to see some real world application where it cause problems before doing so.

tcharding replied:
I added a longer code comment back in.

Kixunil commented on units/src/fee.rs:
```diff
@@ -140,34 +138,33 @@ impl FeeRate {
-        }
-        NumOpResult::Error(E::while_doing(MathOp::Mul))
+    /// Returns [`None`] if overflow occurred.
+    pub fn checked_mul_by_weight(self, weight: Weight) -> Option<Amount> {
```
Well, it's not really meant as a backdoor constructor just as a sensible way to do things. It helps tracking the source better. And yes, we should have it on all relevant types.

Let's not get too attached to consistency with `std`. Returning `Result` instead of `Option` will not be a big deal in practice since those type are nearly identical, the error message would clear it up and it makes sense to provide additional information.

tcharding commented:
I have no idea what to do with your review @Kixunil or with the discussion. I request to merge this as is and address your concerns by way of issues and follow ups.

yancyribbens commented:
> On top of them, I was just asking why are we the only project that uses this unit. If we're the first I think this might be worthy of bitcoin dev ML post or even optech post. I'll review the rest later, I came to GH for a quick look now but should be back soon.

Since this is internal to rust-bitcoin, I don't think external parties are really going to care what the internal semantics are.  The constructors are still sat per kwu and vb so I think really only rust-bitcoin developers are going to care.  I think I saw mention some place of another rust-bitcoin summit taking place, if anything you all could talk about it then if it's still ongoing.

apoelstra commented on units/src/fee.rs:
```diff
@@ -119,8 +119,8 @@ impl FeeRate {
     /// [`NumOpResult::Error`] if an overflow occurred.
```
In 9deab219821a017f07cef1d32b3feef345f91e17:

This doc should be fixed.

tcharding replied:
Added extensive docs, copying some from `checked_mul_by_weight`.

apoelstra commented on units/src/fee_rate/mod.rs:
```diff
@@ -11,26 +11,12 @@ use core::ops;
-    }
-}
-#[doc(inline)]
-pub use encapsulate::FeeRate;
```
Yes, but when changing the internal representation such that the primary conversion functions (to/from sat per kwb) also need to be replaced, the `encapsulate` module isn't clearly a benefit.

If it were my PR I also would've tried to avoid removing the module but I wouldn't be surprised to find that it made the diff bigger and/or required changing the PR ordering significantly.

apoelstra commented:
> I think it should rather work on the internal representation whatever it is.

Github doesn't let me reply to this one, but by definition there is no "get the internal representation whatever it is" method. The point of the `encapsulate` module is that there isn't one.

By the end of the PR everything is put back where it should be, but the individual commits needed to first pull things apart.

@tcharding re "what to do with @Kixunil's review", I believe the comment about missing docs is correct and actionable. So is the "you could have changed the doctest rather than removing it", but I don't know that it's worth updating the PR for the sake of that.

@Kixunil re "we should post on delving", you are welcome to do so! But I can't imagine anybody really cares about feerate rounding errors. We wouldn't either, except that we want a long-term satoshi-precision guarantee of our functions' output *and* we don't want to see any more "Core is more precise than us" type bugs. AFAIK nobody else has both of these goals.

tcharding commented on units/src/result.rs:
````diff
@@ -90,23 +90,6 @@ pub enum NumOpResult<T> {
-    ///     .and_then(|fee| amount.to_signed() - fee);
-    ///
-    /// assert_eq!(amount_after_fee.unwrap(), SignedAmount::from_sat_i32(999_750))
-    /// ```
````
Noted, we can re-visit the docs during final polish of the crate.

tcharding commented:
Changes in force push:

- Added rustdocs as part of https://github.com/rust-bitcoin/rust-bitcoin/pull/4534/commits/1ee57545537dfa050a44254a75318b7aed87e88e
- Added additional patch with code comment

tcharding commented:
Changes in force push are a typo, backticked code in rustdocs incorrectly.

tcharding commented:
Any movement on this one @apoelstra?
